### ISSUE TITLE
[12.0] [auth_ldap] Setting a new password with LDAP enabled allows login with two passwords

### ISSUE THREAD

**Issue by awesome-manuel** [odoo/odoo#35245]
### Documentation states:
_Users' LDAP passwords are never stored in the Odoo database, the LDAP server is queried whenever a user needs to be authenticated. No duplication of the password occurs, and passwords are managed in one place only._

_Odoo does not manage password changes in the LDAP, so any change of password should be conducted by other means in the LDAP directory directly (for LDAP users)._

This is not true, since new passwords are actually stored in the database. Odoo should either implement password change via LDAP (see https://github.com/wwnet/odoo-addons/tree/11.0/auth_ldap_passwd for a possible implementation) or not allow password changes at all.

**Impacted versions:** 12.0

### Steps to reproduce:
- Install and configure LDAP plugin
- Login with credentials from LDAP
- Change your password via Odoo

### Current behavior:
- Odoo stores the new password in it's internal database
- Logout
- Login with your old LDAP password (works)
- Logout
- Login with your new password (works)

### Expected behavior:
- Odoo displays a message that this is not supported
- Logout
- Login with your old LDAP password (works)
- Logout
- Login with your new password (must be rejected)

or

- Odoo updates the LDAP password
- Logout
- Login with your old LDAP password (must be rejected)
- Logout
- Login with your new password (works)

**Comment by Yenthe666** [comment]
@odony I think you're the best guy to respond on this one.

**Comment by odony** [comment]
Yes it's a known limitation of the current LDAP support. Odoo has a pluggable authentication system (like Unix PAM) but it does not have an equivalent PAM for changing passwords. You can authenticate via multiple methods (LDAP, OAuth, password, etc.) yet if you try to change your password in Odoo you will only change the local password one.

The reason is that the system does not know how the user authenticated, so it cannot currently support changing password. We have a backlog task in the R&D pipe to develop this but it requires major changes in the Odoo RPC stack and thus will probably not land in v13 (and certainly not in past stable versions) 

If you find this too confusing for users you might want to entirely disable the password change feature (hide it in the UI) and replace it with a message for users asking them to change their password on the LDAP/AD directly.

It's not really considered a bug either, the multi-authentication PAM system is really a feature.

**Comment by awesome-manuel** [comment]
@odony Actually I don't agree. Having two valid passwords can definitely be considered a bug.

Nevertheless the implementation of variant A above should be fairly straight forward. For the LDAP addon it is quite simple to find out if a user is from LDAP. If LDAP contains an entry for the user AND the user can authenticate against LDAP AND the local password is empty, then the user is from LDAP.

So you only need to use the inverse order for resetting the password: First ask the addon(s) if they can reset the password and if those don't feel responsible for this user reset the local password.

**Comment by pedrobaeza** [comment]
@odony should we close this?

**Comment by odony** [comment]
@pedrobaeza let's not close it yet, we're still planning to improve the situation, perhaps even in 13.0.
Entirely disabling the dual auth mechanism in standard is not desirable now, at least not until we support per-user API keys, or some alternative.

What we could do is a custom version of the password change flow just for LDAP, i.e. if LDAP binding succeeds with the old password, [change it there](https://www.python-ldap.org/en/latest/reference/ldap.html#ldap.LDAPObject.passwd_s), as suggested by @awesome-manuel. We'd still need to decide what to do with the local password if one exists at that point (it could even be the same password). The safest option could be to nullify it always, effectively restoring a single auth password, as the user is trying to reset their credentials anyway, and we should assume the worst.

Users who want to keep two auth methods would still be able to do it by asking an admin to reset their local passwords. Not ideal, but safer.

It would be an improvement over the current mixed behavior, even if this version still suffers from multiple inconsistencies, e.g.:
- Admins are unable to reset the LDAP password of users in Odoo, they can only reset local passwords.
- The password reset mechanism of `auth_signup` would always reset the local password only
- For users with 2 different passwords, changing the local password would work but would not be able to reset the LDAP one.


And then in `master` we work on refactoring the auth layer further, with a better way to identify the auth mechanism that *owns* the session.

**Comment by pedrobaeza** [comment]
OK, thanks for the followup. Let's wait then.

**Comment by azmeuk** [comment]
Things may have changed now with #43101.
